src/contract.rs

```diff
@@ -1,352 +1,47 @@
-pub struct Contract<'a> {
-    hash: H256,
-    module: elements::Module,
-    schedule: &'a Schedule,
-}
-
-impl<'a> Contract<'a> {
-    /// Creates a new Contract from provided code and Schedule.
-    pub fn new<H: ByteHash>(
-        original_code: &[u8],
-        schedule: &'a Schedule,
-    ) -> Result<Self, VMError> {
-        use wasmi_validation::PlainValidator;
-
-        let hash = H256::from_bytes(H::hash(original_code).as_ref());
-
-        let module = elements::deserialize_buffer(original_code)
-            .map_err(|_| VMError::InvalidWASMModule)?;
-
-        // Make sure that the module is valid.
-        wasmi_validation::validate_module::<PlainValidator>(&module)
-            .map_err(|_| VMError::InvalidWASMModule)?;
-
-        let mut contract_module = Contract {
-            hash,
-            module,
-            schedule,
-        };
-
-        contract_module
-            .ensure_no_floating_types()
-            .map_err(|_| VMError::InvalidWASMModule)?;
-        contract_module
-            .ensure_table_size_limit(&schedule)
-            .map_err(|_| VMError::InvalidWASMModule)?;
-
-        contract_module = contract_module
-            .inject_gas_metering()
-            .map_err(|_| VMError::InvalidWASMModule)?
-            .inject_stack_height_metering()
-            .map_err(|_| VMError::InvalidWASMModule)?;
-
-        // Return a `Contract` instance with
-        // __valid__ module.
-        Ok(Contract {
-            hash,
-            module: contract_module.module,
-            schedule,
-        })
-    }
-
-    /// Returns the Contract's hash
-    pub fn hash(&self) -> H256 {
-        self.hash
-    }
-    /// Injects gas metering into the contract
-    fn inject_gas_metering(self) -> Result<Self, failure::Error> {
-        let gas_rules = rules::Set::new(
-            self.schedule.regular_op_cost as u32,
-            Default::default(),
-        )
-        .with_grow_cost(self.schedule.grow_mem_cost as u32)
-        .with_forbidden_floats();
-
-        let contract_module =
-            pwasm_utils::inject_gas_counter(self.module, &gas_rules)
-                .map_err(|_| err_msg("gas instrumentation failed"))?;
-        Ok(Contract {
-            hash: self.hash,
-            module: contract_module,
-            schedule: self.schedule,
-        })
-    }
-
-    /// Injects stack height metering
-    fn inject_stack_height_metering(self) -> Result<Self, failure::Error> {
-        let contract_module = pwasm_utils::stack_height::inject_limiter(
-            self.module,
-            self.schedule.max_stack_height,
-        )
-        .map_err(|_| err_msg("stack height instrumentation failed"))?;
+impl Contract {
+    /// Create a new Contract with initial state and code
+    pub fn new<State, Code, S>(
+        state: State,
+        code: Code,
+        store: &S,
+    ) -> Result<Self, S::Error>
+    where
+        State: Canon<S>,
+        Code: Into<Vec<u8>>,
+        S: Store,
+    {
```

Comment:
Insure compiled was probably premature optimization.

For the other things, we will have to port it again, it will also have to be part of how we do the genesis setup, as per before, to be able to get the contract id before the gas injection code. I will file a new issue on this topic.